### ISSUE TITLE
Nested SharedElements are misplaced in Overlay during transition

### ISSUE THREAD

**Issue by CalamityDeadshot** [mx-albert/compose-shared-elements#3]
Dude, you've done a great job maintaining the original library, thanks for that.  
I've been exploring its capabilities and found a bug.
# The problem
Consider a following use-case: we have a screen with a list of cards and want to animate a clicked card *and* some of its children to a new screen.  
More generally, we want a transition of a **container `SharedElement`** with some other transition of its **children `SharedElement`s**. I found out that this bug occurs when using a `navigation-compose` library, specifically for me it is `androidx.navigation:navigation-compose:2.4.0-alpha10`. Compose versions do not affect this bug, but mine is `1.0.3`.  
Here's some sample code: `NavHost` is itself inside of a `SharedElementsRoot`, switching composables for it.
```kotlin
SharedElementsRoot {
    Surface(
        modifier = Modifier
            .fillMaxSize(),
        color = MaterialTheme.colors.background
    ) {
        val navController = rememberNavController()
        NavHost(
            navController = navController,
            startDestination = "First_screen"
        ) {
            composable(route = "First_screen") {
                LazyColumn(
                    contentPadding = PaddingValues(vertical = 100.dp) // Simulating position in the list
                ) {
                    item {
                        SharedElement(
                            key = "Container",
                            screenKey = "First"
                        ) {
                            SharedCard {
                                navController.navigate("Second_screen")
                            }
                        }
                    }
                }
            }
            composable(route = "Second_screen") {
                SharedElement(
                    key = "Container",
                    screenKey = "Second"
                ) {
                    SharedCardContent {
                        navController.popBackStack()
                    }
                }
            }
        }
    }
}
```
And here is the code for containers:
```kotlin
@Composable
fun SharedCard(
    onClick: () -> Unit
) {
    Card(
        modifier = Modifier
            .fillMaxWidth(),
        elevation = 2.dp,
        shape = RoundedCornerShape(5.dp),
        onClick = onClick
    ) {
        Column(
            modifier = Modifier
                .padding(
                    top = 10.dp,
                    bottom = 8.dp,
                    start = 15.dp,
                    end = 15.dp
                )
        ) {
            Text(
                text = "Card title",
                style = MaterialTheme.typography.h6
            )

            Spacer(modifier = Modifier.height(10.dp))

            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {

                SharedElement(
                    key = "author",
                    screenKey = "First"
                ) {
                    Text("Author name")
                }

                SharedElement(
                    key = "subject",
                    screenKey = "First"
                ) {
                    Text("Subject name")
                }
            }
        }
    }
}

@Composable
fun SharedCardContent(
    onClick: () -> Unit
) {
    Surface(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick),
        color = MaterialTheme.colors.surface,
        elevation = 1.dp
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(20.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp)
        ) {
            SharedElement(
                key = "author",
                screenKey = "Second"
            ) {
                Text("Author: Author name")
            }

            SharedElement(
                key = "subject",
                screenKey = "Second"
            ) {
                Text("Subject: Subject name")
            }

        }
    }
}
```
It looks exactly like we want during transition **forwards**, but fails to position nested `SharedElement`s during transition **backwards**:  
![1](https://user-images.githubusercontent.com/44675043/162847073-4edb2859-4286-46f7-82c4-efcece6e859e.gif)

However, if for some reason the whole tree is recomposed after the *forwards* transition, transition **backwards** plays as expected:  
![2](https://user-images.githubusercontent.com/44675043/162847665-86b650c4-57ba-4629-8843-dc435689549c.gif)  
I assume it's because when `SharedElementsRoot` is recomposed, it considers the `Second` screen to be the `Start` state of the transition and vice versa, so the actual **backwards** transition becomes the **forwards** one, which explains its failure to execute the actual **forwards** transition as intended.  
It would be super cool if you could help fix this issue. Thanks!

**Comment by CalamityDeadshot** [comment]
I experimented further and found out that `SharedElementsRoot` doesn't have to be recomposed to handle backwards transition correctly. It is enough for the `Second` screen to leave the composition and come back, for example by means of navigating to unrelated screen and popping the navigation back stack.  
After that both transitions are executed correctly until another **backwards** one.

**Comment by mxalbert1996** [comment]
Thanks for filing this! Should be fixed by 972c0bf.

**Comment by CalamityDeadshot** [comment]
Some children `SharedElement`s are now inconsistently rendered in the overlay. It was hard to catch and record in test project, but you can see it happening with the last element here:

![1](https://user-images.githubusercontent.com/44675043/163680365-f4cb8e24-7ee5-41ca-beb3-558a63850993.gif)

While testing it inside my main project, I noticed that the same children behave the same way inside items with the same position.  
Increasing duration revealed that they are an fact rendered, but there's some kind of strange state causing them to have the wrong `alpha` value depending on the `fraction` inside `MathUtils.kt/calculateAlpha`
![2](https://user-images.githubusercontent.com/44675043/163681099-ecff339d-4804-4118-9873-195bd52f6ed8.gif)
I would say that the `fadeFraction` in `Placeholder` composable plays the key role here

**Comment by mxalbert1996** [comment]
Please provide a repro so that I can debug.

**Comment by CalamityDeadshot** [comment]
I'm getting a `Failed to resolve: com.mxalbert.sharedelements:shared-elements:0.1.0-SNAPSHOT` error upon building. The needed repo was added to project buildscript:
```
repositories {
    google()
    mavenCentral()
    maven { url 'https://s01.oss.sonatype.org/content/repositories/snapshots' }
}
```
Any ideas?

**Comment by CalamityDeadshot** [comment]
I just copied over the files to make it work. [Here](https://github.com/CalamityDeadshot/shared-elements-debug) is a test repo

**Comment by mxalbert1996** [comment]
Fixed by 2db736a.

**Comment by mxalbert1996** [comment]
Btw, the reason of the unresolved error is that you are putting the repo definition in the `repositories` of `buildscript`, which is wrong. You need to put it in the project's `repositories`. In the project you attached, it's in [setting.gradle](https://github.com/CalamityDeadshot/shared-elements-debug/blob/master/settings.gradle).
